### ISSUE TITLE
as_bindings function not implemented in DigitalRankUpdateCell

### ISSUE THREAD

**Issue by tz-ac** [IBM/aihwkit#216]
## Description

Maybe I am doing something wrong but in my minimal working example a NotImplementedError is raised.
The Error refers to the as_bindings() function of the DigitalRankUpdateCell.

Thanks for you help.

## How to reproduce

Minimal Working Example:
```
from aihwkit.nn import AnalogLinear
from aihwkit.simulator.configs import DigitalRankUpdateRPUConfig
from aihwkit.simulator.configs.devices import ConstantStepDevice, DigitalRankUpdateCell

device = ConstantStepDevice()
rpu = DigitalRankUpdateRPUConfig( DigitalRankUpdateCell(device) )
AnalogLinear(784, 10, rpu_config=rpu)
```
Error: NotImplementedError
in File "C:\ProgramData\Anaconda3\lib\site-packages\aihwkit\simulator\configs\devices.py", line 1151, in as_bindings
raise NotImplementedError

## Other information

* Pytorch version: 1.8.1
* Package version: 0.3.0
* OS: Windows 10 Enterprise Version1909
* Python version: 3.7.1
* Conda version: 4.10.1

**Comment by maljoras** [comment]
Hi @tz-ac ,
thanks for raising the issue! 

The `DigitalRankUpdateCell` is just a base class and cannot be used directly in a `RPUConfig`. Currently, we only have one child of that class, which is the `MixedPrecisionCompound`. So the way you would use that is, e.g.:
```python
rpu_config = DigitalRankUpdateRPUConfig(MixedPrecisionCompound(ConstantStepDevice()))
```

**Comment by tz-ac** [comment]
Hi @maljoras,

thanks for your help. This explains why the error is raised. I will try out your suggestion.

**Comment by maljoras** [comment]
@tz-ac  As a side note, if you are interested in using a perfect update pass (done in floating point), but using the forward and backward pass in analog, you could use the  `IdealDevice`. Note that in this case the forward and backward are still definable as noisy and quantized, but the update will be done in floating point.  
```python
from aihwkit.simulator.configs.devices import IdealDevice
from aihwkit.simulator.configs import SingleRPUConfig
from aihwkit.simulator.configs.utils import UpdateParameters, PulseType

rpu_config = SingleRPUConfig(IdealDevice(), update=UpdateParameters(pulse_type=PulseType.NONE))
```

**Comment by tz-ac** [comment]
@maljoras thank you for this helpful note. 

Your assumption was really close. I am not interested in a perfect update pass, instead I am interested in a reproducable update pass using a `PulsedDevice`. I wanted to remove all sources of variability in a `SingleRPUConfig`.

Using a torch seed, a device with a `construction_seed` and without variability and setting the `UpdateParameters` and `IOParameters` I achieved a reproducable training of a network.
```python
from aihwkit.simulator.configs.devices import ConstantStepDevice
from aihwkit.simulator.configs import SingleRPUConfig
from aihwkit.simulator.configs.utils import UpdateParameters, PulseType, IOParameters

SingleRPUConfig( ConstantStepDevice( w/o variability ),
                 update=UpdateParameters(pulse_type=PulseType.NONE_WITH_DEVICE)),
                 forward=IOParameters(is_perfect=True),
                 backward=IOParameters(is_perfect=True) )
```
Now I only struggle with the problem that a change in certain device parameters has no effect on the network accuracy during and after training. If I cannot figure out the reason (I have the feeling that i should use a different `PulseType`) I will probably open another issue for this in the future. 

Thank you again for your support, I really appreciate it.

**Comment by maljoras** [comment]
@tz-ac yes, best would be to open  new issue with that.  The short answer is that `PulseType.NONE` or `NONE_WITH_DEVICE` skip the pulsed update mechanism altogether. Similar for `is_perfect` for the backward and forward pass (eg skipping the quantization). BEtter would be to explicitly set the noises to zero. Note that currently the noise seed for the cycle-to-cycle variation and the stochastic pulse train generation cannot be fixed and set from the outside. And the internal seed is different from the pytorch seed. Only the construction seed can be set.     

However, you could set all cycle-to-cycle noises to zero and use deterministic pulse trains to make it reproducable:
 
```python
rpu_config = SingleRPUConfig( ConstantStepDevice( dw_min_std=0.0),
                 update=UpdateParameters(pulse_type=PulseType.DETERMINISTIC_IMPLICIT, x_res_implicit=0.01, d_res_implicit=0.01, desired_bl=100)),
                 forward=IOParameters(out_noise=0.0, w_noise=0.0, inp_noise=0.0),
                 backward=IOParameters(out_noise=0.0, w_noise=0.0, inp_noise=0.0) )
```
This would use the pulsed update and should be reproducible.

**Comment by tz-ac** [comment]
@maljoras thank you a lot for all your input. I implemented it and it works!
